File: src/utils/geoDistance.ts
export interface GeoCoordinates {
  latitude: number;
  longitude: number;
}

export interface ProximityConfig {
  radiusKm: number;
  enabled: boolean;
  requireLocationSharing: boolean;
}

export const DEFAULT_PROXIMITY_CONFIG: ProximityConfig = {
  radiusKm: 80,
  enabled: true,
  requireLocationSharing: true,
};

const PROXIMITY_CONFIG_KEY = 'PACT_PROXIMITY_CONFIG';

export function getProximityConfig(): ProximityConfig {
  try {
    const stored = localStorage.getItem(PROXIMITY_CONFIG_KEY);
    if (stored) {
      return { ...DEFAULT_PROXIMITY_CONFIG, ...JSON.parse(stored) };
    }
  } catch (e) {
    console.warn('Failed to load proximity config from localStorage');
  }
  return DEFAULT_PROXIMITY_CONFIG;
}

export function setProximityConfig(config: Partial<ProximityConfig>): void {
  try {
    const current = getProximityConfig();
    const updated = { ...current, ...config };
    localStorage.setItem(PROXIMITY_CONFIG_KEY, JSON.stringify(updated));
  } catch (e) {
    console.warn('Failed to save proximity config to localStorage');
  }
}

export function calculateHaversineDistance(
  point1: GeoCoordinates,
  point2: GeoCoordinates
): number {
  const R = 6371;

  const lat1Rad = toRadians(point1.latitude);
  const lat2Rad = toRadians(point2.latitude);
  const deltaLatRad = toRadians(point2.latitude - point1.latitude);
  const deltaLonRad = toRadians(point2.longitude - point1.longitude);

  const a =
    Math.sin(deltaLatRad / 2) * Math.sin(deltaLatRad / 2) +
    Math.cos(lat1Rad) *
      Math.cos(lat2Rad) *
      Math.sin(deltaLonRad / 2) *
      Math.sin(deltaLonRad / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return R * c;
}

function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}

export function isWithinProximity(
  userLocation: GeoCoordinates,
  siteLocation: GeoCoordinates,
  radiusKm: number = DEFAULT_PROXIMITY_CONFIG.radiusKm
): boolean {
  const distance = calculateHaversineDistance(userLocation, siteLocation);
  return distance <= radiusKm;
}

export function parseGpsCoordinates(
  location: unknown
): GeoCoordinates | null {
  if (!location) return null;

  try {
    let parsed: any = location;
    if (typeof location === 'string') {
      parsed = JSON.parse(location);
    }

    const lat = parsed.latitude ?? parsed.lat;
    const lng = parsed.longitude ?? parsed.lng ?? parsed.lon;

    if (typeof lat === 'number' && typeof lng === 'number' && !isNaN(lat) && !isNaN(lng)) {
      if (lat >= -90 && lat <= 90 && lng >= -180 && lng <= 180) {
        return { latitude: lat, longitude: lng };
      }
    }
  } catch (e) {
    return null;
  }

  return null;
}

export interface SiteProximityCheck {
  isWithinState: boolean;
  isWithinRadius: boolean;
  hasGpsCoordinates: boolean;
  distanceKm: number | null;
  canAccess: boolean;
  reason: string | null;
}

/**
 * Check if a user can access a site based on state + GPS proximity rules.
 * 
 * @param userStateName - The user's assigned state NAME (not ID)
 * @param userLocation - The user's current GPS coordinates
 * @param siteStateId - The site's state ID (optional, prefer siteStateName)
 * @param siteStateName - The site's state NAME
 * @param siteLocation - The site's GPS coordinates (optional)
 * @param config - Proximity configuration (radius, requirements)
 */
export function checkSiteProximity(
  userStateName: string | undefined,
  userLocation: GeoCoordinates | null,
  siteStateId: string | undefined,
  siteStateName: string | undefined,
  siteLocation: GeoCoordinates | null,
  config: ProximityConfig = getProximityConfig()
): SiteProximityCheck {
  const result: SiteProximityCheck = {
    isWithinState: false,
    isWithinRadius: false,
    hasGpsCoordinates: !!siteLocation,
    distanceKm: null,
    canAccess: false,
    reason: null,
  };

  if (!userStateName) {
    result.reason = 'Your profile has no state assigned. Contact your supervisor.';
    return result;
  }

  if (config.requireLocationSharing && !userLocation) {
    result.reason = 'Location sharing must be enabled to view sites. Please enable GPS.';
    return result;
  }

  const normalizedUserState = userStateName.toLowerCase().trim();
  const normalizedSiteState = (siteStateId || siteStateName || '').toLowerCase().trim();
  
  result.isWithinState = 
    normalizedSiteState === normalizedUserState ||
    normalizedSiteState.includes(normalizedUserState) ||
    normalizedUserState.includes(normalizedSiteState);

  if (!result.isWithinState) {
    result.reason = `This site is in a different state.`;
    return result;
  }

  if (!siteLocation) {
    result.canAccess = true;
    result.reason = null;
    return result;
  }

  if (!userLocation) {
    result.canAccess = true;
    result.reason = null;
    return result;
  }

  result.distanceKm = calculateHaversineDistance(userLocation, siteLocation);
  result.isWithinRadius = result.distanceKm <= config.radiusKm;

  if (result.isWithinRadius) {
    result.canAccess = true;
    result.reason = null;
  } else {
    result.canAccess = false;
    result.reason = `This site is ${Math.round(result.distanceKm)} km away, beyond the ${config.radiusKm} km limit.`;
  }

  return result;
}

export function formatDistance(distanceKm: number | null): string {
  if (distanceKm === null) return 'Unknown';
  if (distanceKm < 1) {
    return `${Math.round(distanceKm * 1000)} m`;
  }
  return `${Math.round(distanceKm)} km`;
}

File: mobile_flutter/PACT_mobile-PACT-MAIN-shorebird/lib/utils/geo_distance.dart
import 'dart:math';
import 'package:geolocator/geolocator.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';

class GeoCoordinates {
  final double latitude;
  final double longitude;

  GeoCoordinates({required this.latitude, required this.longitude});

  factory GeoCoordinates.fromJson(Map<String, dynamic> json) {
    final lat = json['latitude'] ?? json['lat'];
    final lng = json['longitude'] ?? json['lng'] ?? json['lon'];
    
    if (lat is num && lng is num) {
      return GeoCoordinates(
        latitude: lat.toDouble(),
        longitude: lng.toDouble(),
      );
    }
    throw FormatException('Invalid coordinates format');
  }

  Map<String, dynamic> toJson() => {
    'latitude': latitude,
    'longitude': longitude,
  };

  bool isValid() {
    return latitude >= -90 && latitude <= 90 && 
           longitude >= -180 && longitude <= 180;
  }
}

class ProximityConfig {
  final double radiusKm;
  final bool enabled;
  final bool requireLocationSharing;

  ProximityConfig({
    this.radiusKm = 80.0,
    this.enabled = true,
    this.requireLocationSharing = true,
  });

  factory ProximityConfig.fromJson(Map<String, dynamic> json) {
    return ProximityConfig(
      radiusKm: (json['radiusKm'] as num?)?.toDouble() ?? 80.0,
      enabled: json['enabled'] as bool? ?? true,
      requireLocationSharing: json['requireLocationSharing'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() => {
    'radiusKm': radiusKm,
    'enabled': enabled,
    'requireLocationSharing': requireLocationSharing,
  };
}

class SiteProximityCheck {
  final bool isWithinState;
  final bool isWithinRadius;
  final bool hasGpsCoordinates;
  final double? distanceKm;
  final bool canAccess;
  final String? reason;

  SiteProximityCheck({
    required this.isWithinState,
    required this.isWithinRadius,
    required this.hasGpsCoordinates,
    this.distanceKm,
    required this.canAccess,
    this.reason,
  });
}

class GeoDistanceUtils {
  static const String _proximityConfigKey = 'PACT_PROXIMITY_CONFIG';
  static const ProximityConfig defaultConfig = ProximityConfig();

  static double calculateHaversineDistance(
    GeoCoordinates point1,
    GeoCoordinates point2,
  ) {
    const double R = 6371; // Earth's radius in km

    final lat1Rad = _toRadians(point1.latitude);
    final lat2Rad = _toRadians(point2.latitude);
    final deltaLatRad = _toRadians(point2.latitude - point1.latitude);
    final deltaLonRad = _toRadians(point2.longitude - point1.longitude);

    final a = sin(deltaLatRad / 2) * sin(deltaLatRad / 2) +
        cos(lat1Rad) * cos(lat2Rad) *
        sin(deltaLonRad / 2) * sin(deltaLonRad / 2);

    final c = 2 * atan2(sqrt(a), sqrt(1 - a));

    return R * c;
  }

  static double _toRadians(double degrees) {
    return degrees * (pi / 180);
  }

  static bool isWithinProximity(
    GeoCoordinates userLocation,
    GeoCoordinates siteLocation, {
    double radiusKm = 80.0,
  }) {
    final distance = calculateHaversineDistance(userLocation, siteLocation);
    return distance <= radiusKm;
  }

  static GeoCoordinates? parseGpsCoordinates(dynamic location) {
    if (location == null) return null;

    try {
      Map<String, dynamic> parsed;
      
      if (location is String) {
        parsed = jsonDecode(location) as Map<String, dynamic>;
      } else if (location is Map) {
        parsed = Map<String, dynamic>.from(location);
      } else {
        return null;
      }

      final coords = GeoCoordinates.fromJson(parsed);
      if (coords.isValid()) {
        return coords;
      }
    } catch (e) {
      return null;
    }

    return null;
  }

  static Future<ProximityConfig> getProximityConfig() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final stored = prefs.getString(_proximityConfigKey);
      if (stored != null) {
        final json = jsonDecode(stored) as Map<String, dynamic>;
        return ProximityConfig.fromJson(json);
      }
    } catch (e) {
      print('Failed to load proximity config: $e');
    }
    return const ProximityConfig();
  }

  static Future<void> setProximityConfig(ProximityConfig config) async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_proximityConfigKey, jsonEncode(config.toJson()));
    } catch (e) {
      print('Failed to save proximity config: $e');
    }
  }

  static SiteProximityCheck checkSiteProximity({
    required String? userStateName,
    required GeoCoordinates? userLocation,
    required String? siteStateName,
    required GeoCoordinates? siteLocation,
    ProximityConfig? config,
  }) {
    final cfg = config ?? const ProximityConfig();

    if (userStateName == null || userStateName.isEmpty) {
      return SiteProximityCheck(
        isWithinState: false,
        isWithinRadius: false,
        hasGpsCoordinates: siteLocation != null,
        canAccess: false,
        reason: 'Your profile has no state assigned. Contact your supervisor.',
      );
    }

    if (cfg.requireLocationSharing && userLocation == null) {
      return SiteProximityCheck(
        isWithinState: false,
        isWithinRadius: false,
        hasGpsCoordinates: siteLocation != null,
        canAccess: false,
        reason: 'Location sharing must be enabled to claim sites. Please enable GPS.',
      );
    }

    final normalizedUserState = userStateName.toLowerCase().trim();
    final normalizedSiteState = (siteStateName ?? '').toLowerCase().trim();

    final isWithinState = normalizedSiteState == normalizedUserState ||
        normalizedSiteState.contains(normalizedUserState) ||
        normalizedUserState.contains(normalizedSiteState);

    if (!isWithinState) {
      return SiteProximityCheck(
        isWithinState: false,
        isWithinRadius: false,
        hasGpsCoordinates: siteLocation != null,
        canAccess: false,
        reason: 'This site is in a different state.',
      );
    }

    if (siteLocation == null) {
      return SiteProximityCheck(
        isWithinState: true,
        isWithinRadius: true,
        hasGpsCoordinates: false,
        canAccess: true,
        reason: null,
      );
    }

    if (userLocation == null) {
      return SiteProximityCheck(
        isWithinState: true,
        isWithinRadius: true,
        hasGpsCoordinates: true,
        canAccess: true,
        reason: null,
      );
    }

    final distanceKm = calculateHaversineDistance(userLocation, siteLocation);
    final isWithinRadius = distanceKm <= cfg.radiusKm;

    if (isWithinRadius) {
      return SiteProximityCheck(
        isWithinState: true,
        isWithinRadius: true,
        hasGpsCoordinates: true,
        distanceKm: distanceKm,
        canAccess: true,
        reason: null,
      );
    } else {
      return SiteProximityCheck(
        isWithinState: true,
        isWithinRadius: false,
        hasGpsCoordinates: true,
        distanceKm: distanceKm,
        canAccess: false,
        reason: 'This site is ${distanceKm.round()} km away, beyond the ${cfg.radiusKm.round()} km limit.',
      );
    }
  }

  static String formatDistance(double? distanceKm) {
    if (distanceKm == null) return 'Unknown';
    if (distanceKm < 1) {
      return '${(distanceKm * 1000).round()} m';
    }
    return '${distanceKm.round()} km';
  }

  static Future<GeoCoordinates?> getCurrentLocation() async {
    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        return null;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          return null;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        return null;
      }

      final position = await Geolocator.getCurrentPosition(
        desiredAccuracy: LocationAccuracy.high,
        timeLimit: const Duration(seconds: 10),
      );

      return GeoCoordinates(
        latitude: position.latitude,
        longitude: position.longitude,
      );
    } catch (e) {
      print('Error getting current location: $e');
      return null;
    }
  }

  static Future<bool> isLocationServiceEnabled() async {
    return await Geolocator.isLocationServiceEnabled();
  }

  static Future<bool> hasLocationPermission() async {
    final permission = await Geolocator.checkPermission();
    return permission == LocationPermission.always || 
           permission == LocationPermission.whileInUse;
  }
}

const ProximityConfig defaultProximityConfig = ProximityConfig();
